segmentation_models_pytorch/unet3plus/decoder.py

```diff
@@ -0,0 +1,245 @@
+            padding=1,
+            use_batchnorm=use_batchnorm,
+        )
+        conv2 = md.Conv2dReLU(
+            out_channels,
+            out_channels,
+            kernel_size=3,
+            padding=1,
+            use_batchnorm=use_batchnorm,
+        )
+        super().__init__(conv1, conv2)
+
+
+class UnetDecoder(nn.Module):
+    def __init__(
+            self,
+            encoder_channels,
+            decoder_channels,
+            n_blocks=5,
+            use_batchnorm=True,
+            attention_type=None,
+            center=True,
+    ):
+        super().__init__()
+
+        """if n_blocks != len(decoder_channels):
+            raise ValueError(
+                "Model depth is {}, but you provide `decoder_channels` for {} blocks.".format(
+                    n_blocks, len(decoder_channels)
+                )
+            )"""
+
+        encoder_channels = encoder_channels[1:]  # remove first skip with same spatial resolution
+        encoder_channels = encoder_channels[::-1]  # reverse channels to start from head of encoder
+
+        # computing blocks input and output channels
+        head_channels = encoder_channels[0]
+        in_channels = encoder_channels[1:]
+        in_channels = in_channels[::-1] #reversing the list
+
+        out_channels = [head_channels] + list(decoder_channels) 
+        previous_channels = [i * n_blocks for i in out_channels] #calculating previos channels i.e concatenated output channels of previous block
+        previous_channels[0] = int(previous_channels[0]/n_blocks) #first tensor as it comes from CenterBlock, hence no concatenation
+
+        if True:
+            self.center = CenterBlock(
+                head_channels, head_channels, use_batchnorm=use_batchnorm
+            )
+        else:
+            self.center = nn.Identity()
+
+        # combine decoder keyword arguments
+        kwargs = dict(use_batchnorm=use_batchnorm, attention_type=attention_type)
+
+        stages = []
+        self.scale = n_blocks-2
+
+        #Blocks for every stage is generated
+        for i in range(n_blocks):
+          stage = DecoderBlock(in_channels, out_channels, previous_channels, self.scale-i, n_blocks, i+1, **kwargs)
+          stages.append(stage)
+  
+        self.stages = nn.ModuleList(stages)
+
+    def forward(self, *features):
+
+        features = features[1:]    # remove first skip with same spatial resolution
+        features = features[::-1]  # reverse channels to start from head of encoder
+
+        head = features[0]
+        current_features = features[1:]
+        current_features = current_features[::-1]
+        current_features = list(current_features)
+        
+        x = self.center(head)  #first tensor is input into centr block
+
+        decoded_features = []
+        decoded_features.append(x)
+        
+        #tensors are calculated for each stage
+        for i, stage in enumerate(self.stages):
+      
+          total_features = current_features.copy()
```

Comment:
Why do we need a copy here?

Reply:
We need a copy here because we don't want to change the values of current_features when we are updating total_features.

Reply:
Not sure if you make a copy of a `list` and tensors will be copied also

Reply:
I have checked that, when I am making the copy of the list consisting of tensors, the tensors are also getting copied.
For e.g., if let say, LIS is a list consisting of 5 tensors of size (64,64). 
LIS_new = LIS.copy()
here, a new list LIS_new will be created which will consist of 5 tensors of size (64,64).

If, I just did LIS_new = LIS, then no new list was getting created, only a new reference for LIS was given. i.e changes done on LIS_new will get reflected in LIS.

But, that is not the case when I do LIS_new = LIS

That is why I have used copy()

If you still think there is some error, let me know. I might be wrong. But based on the experiments and debugging which I did, copy() was working.